### PR TITLE
Implement tags and filtering

### PR DESCRIPTION
**Pull request by kristandre**
#10 Oppgaver kan tagges med vilkårlige tags på denne måten:
```
---
title: Asteroids
level: 4
author: Geir Arne Hjelle
tags:
  tema: [lag spill]
  plattform: [ikke windows, ikke minecraft]
  fag: [engelsk, lesing]
  someTag: [tag1, tag2]
---
```
Alle tags fra alle oppgavene vil dukke opp i filteret på forsiden. 
Antall oppgaver hvert programmeringsspråk har står under ikonet til språket, og det vil oppdateres til å vise antall oppgaver som matcher filteret.

### CODE REVIEW

**Comment by kristandre** [comment]
FrontPage.js ble ganske stor. Den har ansvaret for å hente informasjon om kurs og oppgaver, og lage dette om til javascript objecter som kan sendes videre til andre komponenter. Den har også ansvaret for å oppdatere kurslisten når noe i filteret hukes av. Er det greit at det skjer så mye her, eller burde den deles opp på en måte?

File: src/pages/App.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
 
 const App = React.createClass({
   render() {
     return (
-      <div>
-        <NavBar params={this.props.params}/>
-        <h1>Kodeklubben</h1>
-        {this.props.children}
-      </div>
+      <MuiThemeProvider muiTheme={getMuiTheme()}>
```
Comment (kristandre) [inline src/pages/App.js]: Material-UI bør kanskje kun lastes inn til de komponentene som trenger det?

File: src/util.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export default function capitalize(str) {
```
Comment (NorwegianKiwi) [inline src/util.js]: Jeg foreslår at du bare "export function capitalize", uten default, siden filen heter util.js. Dermed kan vi ha flere funksjoner der etterhvert hvis vi trenger det. Alternativt kan du rename filen til capitalize.js, men synes util.js med flere funksjoner virker som den beste løsningen. Husk at du da må endre import statement i den andre filen til
```
import {capitalize} from '...../util.js';
```

File: src/components/CourseItem.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      this.context.router.push(basePath[1]);
+    }
+  },
+  render() {
+    const course = this.props.course;
+    const name = capitalize(course.name).replace('_', ' ');
+    return (
+      <div style={{display:'inline-block'}} onClick={this.onClick.bind(null, course.path)}>
+        <img src={course.iconPath}
+             style={{width:'80px', margin:'20px'}}/>
```
Comment (NorwegianKiwi) [inline src/components/CourseItem.js]: Forslag: Bruk css-modules på følgende måte.
1) Flytt css ut til en fil, f.eks. `src/components/CourseItem.scss`, som kan se ut som dette
```
.courseItem {
  display: inline-block;
}

.courseLogo {
  width: 80px;
  margin: 20px;
}
```

2) Bytt ut <div> og  sin style med className på følgende vis:
```
import styles from './CourseItem.scss';
...
...
<div className={styles.courseItem} onClick={this.onClick.bind(null, course.path)}>
  <img src={course.iconPath} className={styles.courseLogo} />
    <p>{name}</p>
    <p>Oppgaver: {course.lessons.length}</p>
</div>
```

File: src/components/CourseList.js [hunk 4]
```diff
@@ -1,42 +1,23 @@
...
     );
   }
 });
-//
-// <ul>
-//   <li><NavLink to="/" onlyActiveOnIndex>Front Page</NavLink></li>
-//   <li><NavLink to="/scratch">Scratch</NavLink></li>
-//   <li><NavLink to="/python">Python</NavLink></li>
-// </ul>
+
```
Comment (NorwegianKiwi) [inline src/components/CourseList.js]: Fint om du legger til propTypes på komponenter som bruker props:
```
import React, {PropTypes} from 'react';
...

const CourseList = React.createClass({
...
...
});

CourseList.propTypes = {
  courses: PropTypes.Array
};

export default CourseList;
```

Comment (NorwegianKiwi) [inline src/components/CourseList.js]: Jeg ser jo at du bruker material UI her. Tror du at dere kommer til å bruke dette i stedet for bootstrap, og at vi derfor bør disable / fjerne Bootstrap? Foreløpig er det bare bootstrap sin normalizr som er aktivert, så det gjør jo ingenting om den er der, men hvis det er material ui som skal brukes, så vil vi kanskje på sikt fjerne bootstrap?

File: src/util.js (hunk 2 above)
Comment (NorwegianKiwi) [inline src/util.js]: I tillegg, så er det fint at det lages tester for alt som lages. Nå skal det sies at jeg kanskje ikke helt ser nytten av å skrive tester for html som komponenter rendrer, og komponenter bør være litt "dumme". Men alle funksjoner, som f.eks. capitalize, bør ha tester. I tillegg bør de være "pure", dvs at de ikke endrer argumentene som kommer inn (slik du har gjort med capitalize).

Testene skal ligge under "test", og bør ha samme struktur som koden i "src", bare at filnavnet i test bør ende med `_spec`, så `src/util.js` sine tester bør ligge i `test/util_spec.js`.

Her er et forslag til hvordan denne kan se ut:
```
import { expect } from 'chai';

import {capitalize} from '../src/util';

describe('util.capitalize', () => {

  it('makes first letter in a string capital', () => {
    const text = 'a random string';
    expect(capitalize(text)).to.equal('A random string');
  });

});
```

Du må gjerne legge til flere tester (med flere `it()`), og dersom det kommer flere funksjoner i util.js, så kan du jo legge til flere `describe()` også.

For å kjøre testene, så går du til kommando-linjen og skriver `npm run test`

File: src/pages/FrontPage.js [hunk 5]
```diff
@@ -1,11 +1,214 @@
...
+        list.push(obj);
+      }
+    });
+    return list;
+  },
   render() {
     return (
-      <div>
-        <CourseList/>
+      <div style={{marginTop: '75px'}}>
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Legge css ut i egen fil, slik som jeg beskrev i CourseItem.

File: src/pages/App.js (hunk 1 above)
Comment (NorwegianKiwi) [inline src/pages/App.js]: For å drive med server-side rendering, så må man legge til en userAgent. Kan ikke så mye om dette, men har følgende forslag (sikkert lurere ting å gjøre):
```
<MuiThemeProvider muiTheme={getMuiTheme(null, {userAgent: 'all'})}>
```

Se feks:
http://stackoverflow.com/questions/35481084/react-starter-kit-and-material-ui-useragent-should-be-supplied-in-the-muitheme
http://www.material-ui.com/#/get-started/server-rendering

Comment (NorwegianKiwi) [inline src/pages/App.js]: Når det gjelder spørsmålet om MaterialUI kun skal være i de komponentene som trenger det, så vet jeg ikke nok om MUI ennå. Kom gjerne med argumenter for/imot.

File: src/components/CourseList.js (hunk 4 above)
Comment (NorwegianKiwi) [inline src/components/CourseList.js]: FrontPage.js ble ganske stor, ja. Er nok lurt å dele den opp på et vis. Og helst lage enhetstester til funksjonene den bruker (se forklaringen min på util.js). Skal se om jeg kan komme med litt mer spesifikk tilbakemelding her.

Comment (NorwegianKiwi) [inline src/components/CourseList.js]: Forøvrig, i tilfelle jeg glemmer å si det: 
Dette ser VELDIG kult ut! Bra start! Og bra jobba 👍

File: src/pages/FrontPage.js [hunk 6]
```diff
@@ -1,11 +1,214 @@
...
+  },
+  getTagGroupsFromLessonTags(lessonTagGroups) {
+    const tagGroups = [];
+    /* Tag structure:
+     lessonTagGroups = {
+      TagGroup1: [tagItem1, tagItem2, tagItem3],
+      TagGroup2: [tagItem4, tagItem5],
+      TagGroup3: [tagItem6]
+     } */
+    for (let groupName in lessonTagGroups) {
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Hvis du skal bruke `for ... in ...` på et object, så bør du ha med en sjekk for `hasOwnProperty`.
Et annet forslag er å skrive om og bruke en mer functional programming approach, f.eks.
```
const tagGroups = lessonTagGroups.reduce((result, groupName) => {
  ...
  result.push(.....);
  return result;
}, []);
```

File: src/pages/FrontPage.js [hunk 7]
```diff
@@ -1,11 +1,214 @@
...
+      let tags = lessonTagGroups[groupName];
+
+      // Ignore tagGroups with no tags
+      if (tags == null) continue;
+
+      // Fix non-array tag lists.
+      // This happens if tags is created as string or numbers (e.g. someTagGroupName: tag1, tag2, 12345)
+      // instead of list (e.g. someTagGroupName: [tag1, tag2, 12345]) in YAML
+      if (typeof  tags === 'number') tags = tags.toString();
+      if (typeof tags === 'string') tags = tags.split(/,\s*/);
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Kanskje du kunne skilt ut en egen `fixNonArrayTagList` function? Og så lage en unit-test til denne?

File: src/pages/FrontPage.js [hunk 8]
```diff
@@ -1,11 +1,214 @@
...
+    const lessonTagGroups = lesson.tagGroups;
+    if(lessonTagGroups.length == 0 && filterTagGroup !== null) return false;
+    for(var i = 0; i < lessonTagGroups.length; i++){
+      const lessonTagGroup = lessonTagGroups[i];
+      //TODO: Clean
+      if(lessonTagGroup.name.toLowerCase() == filterTagGroup.name.toLowerCase() && this.isSubArray(filterTagGroup.tags, lessonTagGroup.tags))return true;
+    }
+    return false;
+  },
+  mergeTagGroups(tagGroupA, tagGroupB) {
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Denne funksjonen kan du skrive litt enklere:
```
tagGroupA.tags = [...new Set(tagGroupA.tags.concat(tagGroupB.tags))];
```

Her bruker jeg trikset med at
```
const uniqueArray  = [ ...new Set(array) ]
```

Når det er sagt, så prøv å unngå funksjoner som endrer argumentene, altså prøv å lage "pure" functions, uten side effects. Mye lettere å teste, mye mindre sjanse for feil.

File: src/pages/FrontPage.js [hunk 9]
```diff
@@ -1,11 +1,214 @@
...
+    tagGroupB.tags.forEach(function (tag) {
+      // Add the new tag to the existing tagGroup if it does not already have it
+      if (tagGroupA.tags.indexOf(tag) < 0)tagGroupA.tags.push(tag);
+    });
+  },
+  sortCourses(a, b) {
+    return b.lessons.length - a.lessons.length;
+  },
+  isSubArray(subArr, arr) {
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Hva er det egentlig denne funksjonen gjør?

File: src/pages/FrontPage.js [hunk 10]
```diff
@@ -1,11 +1,214 @@
...
 const FrontPage = React.createClass({
+  getInitialState() {
+    return {
+      filter: [],
+      tagGroups: [],
+      allCourses: [],
+      filteredCourses: []
+    };
+  },
+  componentWillMount() {
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Jeg vet det er fristende å lage `courses` og `tagGroups` i samme funksjon, siden du bruker litt av den samme infoen når du gjøre paths.forEach(), og du tenker at det er sikkert raskere å lage begge samtidig.

Min anbefaling er at du unngår denne fristelsen, for dette er optimalisering på bekostning av kompleksitet, og selv om denne optimaliseringen kan virke lur, så blir altså koden mer kompleks, og hvis du skiller ut funksjonaliteten her i to funksjoner getCourses() og getTagGroups() (som kanskje begge tar et `paths`-argument?), så vil du kanskje se at det ikke blir så ille likevel? Det gjør det ihvertfall lettere å forstå koden, og det blir lettere å skrive enhetstester.

Forstod du hva jeg mente nå, eller var jeg litt tåkete?

File: src/pages/FrontPage.js [hunk 11]
```diff
@@ -1,11 +1,214 @@
...
+      // Ignore tagGroups with no tags
+      if (tags == null) continue;
+
+      // Fix non-array tag lists.
+      // This happens if tags is created as string or numbers (e.g. someTagGroupName: tag1, tag2, 12345)
+      // instead of list (e.g. someTagGroupName: [tag1, tag2, 12345]) in YAML
+      if (typeof  tags === 'number') tags = tags.toString();
+      if (typeof tags === 'string') tags = tags.split(/,\s*/);
+
+      tagGroups.push({
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Hvorfor beholder du ikke strukturen til lessonTagGroups, der groupName (en string) er key og tags (en array) er verdien? Det virker jo mye lettere å kombinere tags i etterkant da, slik du gjør med getListWithDistinctObjects?

File: src/components/filter/LessonFilter.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      <Card>
+        <CardTitle title="Filter" />
+        <CardText>
+          {tagGroupNodes}
+        </CardText>
+      </Card>
+    );
+  }
+});
+
```
Comment (NorwegianKiwi) [inline src/components/filter/LessonFilter.js]:
```
LessonFilter.propTypes = {
  ...
}
```

File: src/components/filter/LessonFilter.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import FilterGroup from './FilterGroup';
+import Divider from 'material-ui/Divider';
+import {Card, CardText, CardTitle} from 'material-ui/Card';
+
+const LessonFilter = React.createClass({
+  render(){
+    const tagGroups = this.props.tagGroups;
+    const tagGroupNodes = tagGroups.map((tagGroup, idx) => {
```
Comment (NorwegianKiwi) [inline src/components/filter/LessonFilter.js]: Hvis du bestemmer deg for å la tagGroups beholde strukturen til lessonTagGroups, så kan du jo bruke
```
const tagGroupNodes = Object.keys(tagGroups).map((tagGroup, idx) => {
  ...
});

File: src/components/filter/FilterGroup.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    });
+    return (
+      <div>
+        <p>{groupName}</p>
+        {filterItems}
+      </div>
+    );
+  }
+});
+
```
Comment (NorwegianKiwi) [inline src/components/filter/FilterGroup.js]:
```
FilterGroup.propTypes = {
  ...
}
```

File: src/components/filter/FilterItem.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    return (
+      <Checkbox
+        label={tag}
+        checked={this.state.checked}
+        onCheck={this.onCheck}
+      />
+    );
+  }
+});
+
```
Comment (NorwegianKiwi) [inline src/components/filter/FilterItem.js]:
```
FilterItem.propTypes = {
  ...
}
```

File: src/components/filter/FilterItem.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    this.setState({checked: !checked});
+    this.props.onCheck(this.props.tag, !checked);
+  },
+  render(){
+    const tag = capitalize(this.props.tag);
+    return (
+      <Checkbox
+        label={tag}
+        checked={this.state.checked}
+        onCheck={this.onCheck}
```
Comment (NorwegianKiwi) [inline src/components/filter/FilterItem.js]: Hm, jeg ser at det er FilterItem som har staten for om et filter er checked eller ei, men at onCheck er sendt ned fra FrontPage (handleOnCheck) via flere lag med komponenter for å gi beskjed opp til FrontPage slik at den kan displaye riktige kurs. Allerede her begynner det å føles riktig å innføre Redux... men kan jo tenke litt på om det er en enklere måte å gjøre det på. Men en ting om gangen. Hvis vi skal legge inn Redux, gjør vi det i en egen pullrequest senere. La oss fikse de andre tingene først :)

File: src/pages/FrontPage.js [hunk 17]
```diff
@@ -1,11 +1,214 @@
 import React from 'react';
 import CourseList from '../components/CourseList';
+import LessonFilter from '../components/filter/LessonFilter';
+
+const iconContext = require.context('lessonSrc/', true, /^\.\/[^\/]*\/logo-black\.png/);
+const lessonContext = require.context('onlyFrontmatter!lessonSrc/', true,
+  /^\.\/[^\/]*\/[^\/]*\/(?!README\.md$)[^\/]*\.md/);
 
 const FrontPage = React.createClass({
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Jeg må innrømme at, etter å ha sett på dette i noen timer nå, ennå ikke helt har oversikt over detaljene i hvordan disse tag-strukturene henger sammen. Jeg har på følelsen av at en del av denne koden kan forenkles, men da må jeg forstå litt bedre hva du har tenkt. Kunne du skrevet en litt mer utførlig kommentar (i koden) som forklarer dette i litt mer detalj? Det gjør det også enklere for folk som skal vedlikeholde koden senere. (Alternativt, hvis du klarer å forenkle koden så mye at det er opplagt hva koden gjør, og/eller legger ved nok unit-tests, så kan du krympe inn kommentarene.)

**Comment by arve0** [comment]
Trenger vi en slik struktur på tags? Hadde det holdt med en liste av strings? Eventuell funksjonalitet kan leite etter 'windows' fra en liste. Less is more?

**Comment by kristandre** [comment]
Tenker du på å droppe gruppering av forskjellige typer tags?
Så det blir 
`tags: [windows, fysikk, lag spill, tag1, tag2, ...]`

**Comment by arve0** [comment]
Det var tanken. Skal litt til for å få navnekræsj?

**Comment by kristandre** [comment]
Navnekræsj skal det nok litt til for å få, ja.
Jeg la til grupper etter denne beskrivelsen av hvordan filtrering skal fungere:
> Dette er i praksis checkboxes (eller knapper som holder seg inne). De er gruppert i tema ("Lag spill", "Lag app", "Roboter"), plattform ("Bare på iPad", "Ikke Windows", "Ikke Minecraft"), og fag ("Regning", "Engelsk", "Fysikk", "Lesing").

Hvis vi kun har èn liste med tags, så må vi forhåndsdefinere at 'lag spill' hører til tema, 'ikke Minecraft' hører til plattform osv. Eller er det noe jeg har oversett?
Jeg tenkte at det var greit med en fleksibel løsning der filtreringsmenyen blir generert med grupper utifra oppgavene. Da trenger man ikke endre i koden hver gang man skal tillate en ny tag eller legge til/endre en gruppe. Også hvis vi skal ha støtte for flere språk, så er det jo greit at andre land kan lage sine egne tags. Ulempen er jo at man må huske hva de forskjellige gruppene heter når man lager tags.

**Comment by NorwegianKiwi** [comment]
Jeg synes @kristandre har et godt poeng her. Spesielt hvis noen av tagGroups'ene ikke skal sees av elever, som f.eks. fag. Da blir det lettere å skjule disse når man er i elev-modus. Dessuten tror jeg ikke det trenger være spesielt mer komplisert enn alle i én haug --- som nevnt, objektet du får tilbake fra frontmatter er jo allerede et object med tagGroupName som key og array av tags som verdi.

File: src/pages/FrontPage.js [hunk 18]
```diff
@@ -1,11 +1,170 @@
...
+      return this.lessonHasTags(lesson, filter);
+    });
+  },
+  lessonHasTags(lesson, filter) {
+    // Filter is empty
+    if(Object.keys(filter).length === 0) return true;
+
+    const lessonTags = lesson.tags;
+    for(let groupName in filter){
+      if(!lessonTags.hasOwnProperty(groupName))return false;
```
Comment (NorwegianKiwi) [inline src/pages/FrontPage.js]: Jeg har ikke sett så nøye på dette ennå, men bare ser i forbifarten at du gjør lessonTags.hasOwnProperty()... burde det ikke vært filter.hasOwnProperty, siden det er denne du looper over?

Comment (kristandre) [inline src/pages/FrontPage.js]: Skal være begge!

File: test/util_spec.js [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+
+  describe('capitalize', () => {
+
+    it('makes first letter in a string capital', () => {
+      const text = 'a random string';
+
+      expect(capitalize(text)).to.equal('A random string');
+    });
+
+    it('is pure', () => {
```
Comment (NorwegianKiwi) [inline test/util_spec.js]: Først må jeg bare si at jeg er kjempeimponert over hvor fort du plukker opp kommentarene og forbedrer koden!

Og veldig kult at du legger inn disse "is pure"-testene! Jeg har et lite tips her, for å spare litt arbeid, og det går ut på at det finnes en måte vi kan teste at funksjonene er pure samtidig som vi gjør de andre testene.

Argumenter til funksjoner kan sendes inn "by value" eller "by reference". Når vi sender inn "by value", så er det ikke mulig å endre på original-verdiene, så da trenger man ikke være bekymret for at funksjonen skal være "un-pure". Dette gjelder tall og booleans.

Når man sender inn objekter og arrays (som er "by reference"), så kan man bruke Object.freeze() (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) for å fryse dem før man sender dem inn; hvis de da blir endret, så får man beskjed. I praksis er det bedre å bruke f.eks. deep-freeze (https://github.com/substack/deep-freeze), som rekursivt går gjennom og fryser objekter og arrays.

Når det gjelder strings, så er disse gjort immutable i javascript, så de oppfører seg som om de ble sendt "by value"; derfor trenger man heller ikke fryse disse.

Dette betyr at vi ikke trenger "is pure" testene; når man sender inn strings, tall og booleans er det slik automatisk, og man kan bruke deep-freeze før man sender inn arrays og objects.

Se forøvrig
http://docstore.mik.ua/orelly/webprog/jscript/ch11_02.htm

Comment (kristandre) [inline test/util_spec.js]: Takk for tipset, dette sparer mye arbeid!

File: test/util_spec.js [hunk 20]
```diff
@@ -0,0 +1,341 @@
...
+
+      expect(capitalize(text)).to.equal('A random string');
+    });
+  });
+
+  describe('getTags', () => {
+
+  });
+
+  describe('getCourses', () => {
```
Comment (kristandre) [inline test/util_spec.js]: Jeg sliter litt med å finne ut hvordan getTags og getCourses kan testes, fordi de er avhengige av hva som finnes i ../oppgaver. Finnes det en måte å lage en slags 'mock' for ../oppgaver?

File: package.json [hunk 21]
```diff
@@ -40,6 +40,7 @@
     ]
   },
   "dependencies": {
+    "material-ui": "^0.15.0",
```
Comment (NorwegianKiwi) [inline package.json]: Når jeg gjøre "npm install" så får jeg følgende beskjed:
```
npm WARN material-ui@0.15.1 requires a peer of react-tap-event-plugin@^1.0.0 but none was installed.
```

Mistenker at
```
npm install --save react-tap-event-plugin
```
løser saken.
(Jeg forstod det slik at material-ui etterhvert ikke vil ha denne dependency'en, men enn så lenge...)

File: src/util.js [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+/**
+ * Add tag to filter
+ * @param {Object} tag
+ * @param {Object} filter
+ * @returns {Object} filter with tag merged into it
+ */
+export function addTagToFilter(tag, filter) {
+  return mergeObjects(filter, tag);
```
Comment (NorwegianKiwi) [inline src/util.js]: Bitte liten filleting, ikke viktig: Så vidt jeg kan se er mergeObjects symmetrisk mhp argument nr 1 og nr 2. Så da kan noen som leser koden begynne å lure på hvorfor addTagToFilter får inn (tag, filter) og sender det videre med (filter, tag). Som sagt, ikke viktig.

Det at du har valgt å beholde addTagToFilter(), og ikke kalle mergeObjects direkte, regner jeg med er fordi du tenker at du kanskje senere kommer til å utvide addTagToFilter()? Eller det gjør koden mer lesbar? Et evt. motargument (og vi er fortsatt på pirkestadiet her, ikke veldig viktig) er at "addTagToFilter" høres ikke symmetrisk ut med tanke på arg1 og arg2, men det gjør mergeObjects. Så ved å bruke mergeObjects direkte, så forstår leseren bedre hva som skjer med objektene. Fordelen med å beholde addTagToFilter er at du er tydeligere på hva intensjonen er (men gjemmer "implementasjonen"). Er vel bare smakssak-greier det her.

Uansett, du trenger ikke endre noe her om du ikke vil, bare litt ekstra babbel fra meg.

Comment (NorwegianKiwi) [inline src/util.js]: Vel, et argument til for å bruke bare mergeObjects, er at du slipper unna med én test mindre, da... hehe

File: test/util_spec.js [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+      deepFreeze(tag);
+      deepFreeze(filter);
+      expect(addTagToFilter(tag, filter)).to.eql({
+        platform: ['browser', 'windows'],
+        subject: ['physics']
+      });
+    });
+  });
+
+  describe('removeTagFromFilter', () => {
```
Comment (NorwegianKiwi) [inline test/util_spec.js]: Denne testen er kanskje litt "snill". Litt flere "corner-cases" hadde vært fint. Enten en litt mer komplisert test, eller legge til et par flere "it()". F.eks. hva om du forsøker å fjerne en tag som ikke finnes? Hva om tag={}? Hva om tag inneholder en groupName som ikke finnes i filter? Hva om tag inneholder mer enn 1 groupName?

File: src/util.js [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+
+/**
+ * Get courses that has at least one lesson that matches filter
+ * @param {Array} courses
+ * @param {Object} filter
+ * @returns {Array} filtered courses
+ */
+export function filterCourses(courses, filter) {
+  const coursesWithFilteredLessons = courses.map(course => {
+    const newCourse = Object.assign({}, course);
```
Comment (NorwegianKiwi) [inline src/util.js]: Tips: Siden vi bruker ES6 stage2, så kan du faktisk skrive dette som
```
const newCourse = {...course};
```

Et mer komplekst eksempel er
```
const newObj = Object.assign({}, oldObj, {mykey1: myvalue1, mykey2:myvalu2});
```
blir til
```
const newObj = {...oldObj, mykey1: myvalue1, mykey2:myvalu2};
```

File: src/util.js [hunk 25]
```diff
@@ -0,0 +1,193 @@
...
+  return tagItems;
+}
+
+/**
+ * Check if lesson contains all tags in filterTags
+ * @param {Object} lesson
+ * @param {Object} filterTags
+ * @returns {boolean} lesson contains all tags required by filter
+ */
+export function lessonHasTags(lesson, filterTags) {
```
Comment (NorwegianKiwi) [inline src/util.js]: Bare et forslag: Er du enig i at `lessonHasTags` egentlig burde hete `lessonHasAllTags`?

File: src/util.js [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+  // Filter is empty
+  if(Object.keys(filterTags).length === 0) return true;
```
Comment (NorwegianKiwi) [inline src/util.js]: Burde ikke denne returnert false?
Og kanskje lage en test for dette tilfellet (at filterTags.length === 0)?

File: src/util.js [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+
+  const lessonTags = lesson.tags;
+  for(let groupName in filterTags){
+    if(!lessonTags.hasOwnProperty(groupName) || !filterTags.hasOwnProperty(groupName))return false;
```
Comment (NorwegianKiwi) [inline src/util.js]: Dette tror jeg blir feil.

Se på følgende eksempel (testes f.eks. på jsfiddle.net, åpne inspector og se i console):
```
Object.prototype.hi=function(){ console.log('hi'); }
const buz={ foo: 'bar' };

for (let k in buz) {
	if (buz.hasOwnProperty(k)) {
  	console.log('hasOwn: ' + k);
  } else {
    console.log('has not: ' + k);
  }
}
```
Så problem med koden er at dersom noen har, et annet sted i koden, utvidet Object.prototype med f.eks. Object.prototype.hi, så vil denne funksjonen returnere false om den kommer til 'hi', selv om det ikke er en tag. (Altså, groupName blir 'hi', men hasOwnProperty blir false for både lessonTags og filterTags).

Kanskje det er lettere å bruke map, filter eller reduce? (Vet ikke, bare spør)
(map, filter og reduce gjør "hasOwnProperty" automagisk).

File: src/util.js (hunk 22 above)
Comment (kristandre) [inline src/util.js]: Tanken var å gjøre koden mer lesbar. 

`const filter = checked ? addTagToFilter(tag, this.state.filter) : removeTagFromFilter(tag, this.state.filter);`

Jeg har prøvd å holde FrontPage så liten og lesbar som mulig, slik at man raskt får oversikt over hva som skjer der. Alle implementasjonsdetaljer er flyttet til util.js.
Skal selvfølgelig endre til mergeObjects hvis det gjør koden enklere å forstå

File: test/util_spec.js (hunk 23 above)
Comment (kristandre) [inline test/util_spec.js]: Skal lage noen flere tester her

File: src/util.js (hunk 25 above)
Comment (kristandre) [inline src/util.js]: Enig!

File: src/util.js (hunk 27 above)
Comment (kristandre) [inline src/util.js]: Dette visste jeg ikke at kunne skje. Jeg skal prøve å skrive den om uten å bruke 'for...in'

Comment (NorwegianKiwi) [inline src/util.js]: Kanskje bare
```
export function lessonHasTags(lesson, filterTags) {
  // Filter is empty
  if(Object.keys(filterTags).length === 0) return true;

  const lessonTags = lesson.tags;
  for(let groupName in filterTags){
    if(filterTags.hasOwnProperty(groupName)) {
      const filterTagItems = filterTags[groupName];
      const lessonTagItems = lessonTags[groupName] || [];
      // Check if there exist at least one filterTag that the lesson does not have
      if (filterTagItems.find(tagItem => lessonTagItems.indexOf(tagItem) < 0)) return false;
    }
  }
  // Lesson contains all tags in the filter
  return true;
}
```
?

File: src/util.js [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+
+/**
+ * Get courses that has at least one lesson that matches filter
+ * @param {Array} courses
+ * @param {Object} filter
+ * @returns {Array} filtered courses
+ */
+export function filterCourses(courses, filter) {
+  const coursesWithFilteredLessons = courses.map(course => {
+    const newCourse = {...course}
```
Comment (NorwegianKiwi) [inline src/util.js]: Mangler semicolon.
Kjør gjerne eslint før du committer, så finner den sånt for deg:
```
npm run eslint -s
```

File: src/components/CourseList.js [hunk 29]
```diff
@@ -1,42 +1,37 @@
...
-      this.context.router.push(basePath[1]);
-    }
+  sortCourses(a, b) {
+    return b.lessons.length - a.lessons.length;
   },
   render() {
-    const icons = iconContext.keys().map((path, idx) =>{
+    const sortedCourses = this.props.courses.sort(this.sortCourses);
+    const courses = sortedCourses.map((course, idx) => {
+      if (course.lessons.length == 0) return null;
```
Comment (NorwegianKiwi) [inline src/components/CourseList.js]: Burde være trippellik: ===

File: src/components/filter/LessonFilter.js [hunk 30]
```diff
@@ -0,0 +1,35 @@
...
+        <CardText>
+          {tagNodes}
+        </CardText>
+      </Card>
+    );
+  }
+});
+
+LessonFilter.propTypes = {
+  tagGroups: PropTypes.object,
```
Comment (NorwegianKiwi) [inline src/components/filter/LessonFilter.js]: tagGroups har skiftet navn til tags.

File: src/components/filter/FilterGroup.js [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+FilterGroup.propTypes = {
+  tagGroup: PropTypes.string,
```
Comment (NorwegianKiwi) [inline src/components/filter/FilterGroup.js]: tagGroup er byttet ut med groupName

File: test/util_spec.js (hunk 20 above)
Comment (NorwegianKiwi) [inline test/util_spec.js]: Vi kan jo bare lage en fake lessonContext og iconContext som vi sender inn med testene. Hvis de blir litt store, så kan man legge dette som consts i egne filer. Men jeg foreslår at vi legger dette som issues, så kan vi ta det i en annen PR.

**Comment by NorwegianKiwi** [comment]
Fant kanskje en bug også: Dersom jeg lager en oppgave med en tom tag-liste (f.eks. plattform: []), så blir det en checkbox uten label. Er dette greit?

File: test/util_spec.js [hunk 32]
```diff
@@ -0,0 +1,476 @@
...
+
+      deepFreeze(tag);
+      deepFreeze(filter);
+      expect(removeTagFromFilter(tag, filter)).to.eql({
+        platform: ['windows', 'browser'],
+        subject: ['physics']
+      });
+    });
+
+    it('removes nothing if tag contains more than zero tagItems', () => {
```
Comment (NorwegianKiwi) [inline test/util_spec.js]: Skulle det vært 'removes nothing if tag contains zero tagItems'?
